# 1
#Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

missing_values = df.isna().sum()

print("Missing values in each column:")
print(missing_values)

Missing values in each column:
row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


Yes, there is a missing value in this set of data

# 2
#Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a pandas DataFrame has, and then

In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)

num_rows, num_columns = villagers_df.shape

print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_columns}")

Number of rows: 391
Number of columns: 11


This set of data has a total of 391 rows and 11 columns.

Observation: In this dataset, an observation is a row of information about a character, with each row providing details about a character, such as its name, species, and personality.

Variables: Variables are the types of information we have for each character. These are the columns in the dataset, such as name, species, and personality. Each column represents a specific type of detail about the role

# 3
Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

In [1]:
import pandas as pd

data = {
    'Age': [25, 32, 40, 29, 33, 36, 28, 30],
    'City': ['New York', 'Paris', 'New York', 'London', 'Paris', 'London', 'New York', 'London'],
    'Salary': [50000, 54000, 61000, 45000, 62000, 58000, 52000, 51000]
}

df = pd.DataFrame(data)

In [2]:
df.describe()

Age        Salary
count   8.00000      8.000000
mean   31.62500  54125.000000
std     4.74906   5841.660723
min    25.00000  45000.000000
25%    28.75000  50750.000000
50%    31.00000  53000.000000
75%    33.75000  58750.000000
max    40.00000  62000.000000

The describe() method provides a statistical summary (such as count, mean, standard deviation, minimum and maximum values) of the numeric columns in the dataset.

In [3]:
df['City'].value_counts()

City
New York    3
London      3
Paris       2
Name: count, dtype: int64

This will give you the frequency of each unique value in the "City" column.

# 4
#If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

df.shape reports the total number of columns (numeric and non-numeric).
df.describe() only reports numeric columns unless you explicitly request non-numeric analysis

The "count" column in df.describe() reflects the number of non-missing values in each numeric column. If the column contains missing values, the count will be lower than the total number of rows.

# 5
#Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference


df.shape: Properties or characteristics of an object. It stores static information and can be accessed without parentheses.
df.describe(): A function associated with an object that performs an operation or calculation. It is called with parentheses.

In [2]:
import pandas as pd 

In [2]:
##url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"

In [3]:
local = "/home/jovyan/villagers.csv"

In [4]:
df = pd.read_csv(local)

In [5]:
df

row_n        id      name  gender    species birthday personality  \
0       2   admiral   Admiral     NaN       bird     1-27      cranky   
1       3   agent-s   Agent S  female   squirrel      7-2       peppy   
2       4     agnes     Agnes     NaN        pig     4-21        uchi   
3       6        al        Al    male    gorilla    10-18        lazy   
4       7   alfonso   Alfonso    male  alligator      6-9        lazy   
5       8     alice     Alice     NaN      koala     8-19      normal   
6       9      alli      Alli     NaN  alligator     11-8      snooty   
7      10    amelia    Amelia  female      eagle    11-19      snooty   
8      11  anabelle  Anabelle  female   anteater     2-16       peppy   
9      13   anchovy   Anchovy    male       bird      3-4        lazy   
10     14     angus     Angus    male       bull     4-30      cranky   
11     15  anicotti  Anicotti  female      mouse     2-24       peppy   
12     16     ankha     Ankha  female        cat     9-22      snooty   
13     17  annalisa  Annalisa  female   anteater      2-6      normal   
14     18  annalise  Annalise  female      horse     12-2      snooty   
15     19   antonio   Antonio    male   anteater    10-20        jock   
16     20    apollo    Apollo    male      eagle      7-4      cranky   
17     21     apple     Apple  female    hamster     9-24       peppy   
18     22    astrid    Astrid  female        NaN      9-8      snooty   
19     23     audie     Audie  female       wolf     8-31       peppy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
5       Surfin' K.K.    guvnor     villager-alice   
6       K.K. Chorale   graaagh      villager-alli   
7        K.K. Condor    eaglet    villager-amelia   
8         Aloha K.K.    snorty  villager-anabelle   
9       K.K. Ragtime    chuurp   villager-anchovy   
10        Senor K.K.    macmoo     villager-angus   
11        Neapolitan   cannoli  villager-anicotti   
12  Marine Song 2001   me meow     villager-ankha   
13         King K.K.   gumdrop  villager-annalisa   
14       K.K. Sonata    nipper  villager-annalise   
15        K.K. Faire      honk   villager-antonio   
16         K.K. Rock       pah    villager-apollo   
17        I Love You  cheekers     villager-apple   
18        I Love You    my pet    villager-astrid   
19               NaN   foxtrot     villager-audie   

                                                  url  
0   https://villagerdb.com/images/villagers/thumb/...  
1                                                 NaN  
2   https://villagerdb.com/images/villagers/thumb/...  
3   https://villagerdb.com/images/villagers/thumb/...  
4   https://villagerdb.com/images/villagers/thumb/...  
5   https://villagerdb.com/images/villagers/thumb/...  
6   https://villagerdb.com/images/villagers/thumb/...  
7   https://villagerdb.com/images/villagers/thumb/...  
8   https://villagerdb.com/images/villagers/thumb/...  
9   https://villagerdb.com/images/villagers/thumb/...  
10  https://villagerdb.com/images/villagers/thumb/...  
11  https://villagerdb.com/images/villagers/thumb/...  
12  https://villagerdb.com/images/villagers/thumb/...  
13  https://villagerdb.com/images/villagers/thumb/...  
14  https://villagerdb.com/images/villagers/thumb/...  
15  https://villagerdb.com/images/villagers/thumb/...  
16  https://villagerdb.com/images/villagers/thumb/...  
17  https://villagerdb.com/images/villagers/thumb/...  
18  https://villagerdb.com/images/villagers/thumb/...  
19  https://villagerdb.com/images/villagers/thumb/...

In [6]:
df.describe()

row_n
count  20.000000
mean   12.900000
std     6.592659
min     2.000000
25%     7.750000
50%    13.500000
75%    18.250000
max    23.000000

In [7]:
df['id'].describe()

count          20
unique         20
top       admiral
freq            1
Name: id, dtype: object

In [8]:
del df['gender']

In [9]:
df.dropna()

row_n        id      name    species birthday personality  \
0       2   admiral   Admiral       bird     1-27      cranky   
2       4     agnes     Agnes        pig     4-21        uchi   
3       6        al        Al    gorilla    10-18        lazy   
4       7   alfonso   Alfonso  alligator      6-9        lazy   
5       8     alice     Alice      koala     8-19      normal   
6       9      alli      Alli  alligator     11-8      snooty   
7      10    amelia    Amelia      eagle    11-19      snooty   
8      11  anabelle  Anabelle   anteater     2-16       peppy   
9      13   anchovy   Anchovy       bird      3-4        lazy   
10     14     angus     Angus       bull     4-30      cranky   
11     15  anicotti  Anicotti      mouse     2-24       peppy   
12     16     ankha     Ankha        cat     9-22      snooty   
13     17  annalisa  Annalisa   anteater      2-6      normal   
14     18  annalise  Annalise      horse     12-2      snooty   
15     19   antonio   Antonio   anteater    10-20        jock   
16     20    apollo    Apollo      eagle      7-4      cranky   
17     21     apple     Apple    hamster     9-24       peppy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
5       Surfin' K.K.    guvnor     villager-alice   
6       K.K. Chorale   graaagh      villager-alli   
7        K.K. Condor    eaglet    villager-amelia   
8         Aloha K.K.    snorty  villager-anabelle   
9       K.K. Ragtime    chuurp   villager-anchovy   
10        Senor K.K.    macmoo     villager-angus   
11        Neapolitan   cannoli  villager-anicotti   
12  Marine Song 2001   me meow     villager-ankha   
13         King K.K.   gumdrop  villager-annalisa   
14       K.K. Sonata    nipper  villager-annalise   
15        K.K. Faire      honk   villager-antonio   
16         K.K. Rock       pah    villager-apollo   
17        I Love You  cheekers     villager-apple   

                                                  url  
0   https://villagerdb.com/images/villagers/thumb/...  
2   https://villagerdb.com/images/villagers/thumb/...  
3   https://villagerdb.com/images/villagers/thumb/...  
4   https://villagerdb.com/images/villagers/thumb/...  
5   https://villagerdb.com/images/villagers/thumb/...  
6   https://villagerdb.com/images/villagers/thumb/...  
7   https://villagerdb.com/images/villagers/thumb/...  
8   https://villagerdb.com/images/villagers/thumb/...  
9   https://villagerdb.com/images/villagers/thumb/...  
10  https://villagerdb.com/images/villagers/thumb/...  
11  https://villagerdb.com/images/villagers/thumb/...  
12  https://villagerdb.com/images/villagers/thumb/...  
13  https://villagerdb.com/images/villagers/thumb/...  
14  https://villagerdb.com/images/villagers/thumb/...  
15  https://villagerdb.com/images/villagers/thumb/...  
16  https://villagerdb.com/images/villagers/thumb/...  
17  https://villagerdb.com/images/villagers/thumb/...

# 6

if colum is numeric like row_n, the output might look like this:

count  20.000000   The number of non-null (non-missing) entries in the column.

mean   12.900000   The average of the non-null values in the column.

std    6.592659    A measure of how different or spread out the values in a column are from the mean.A higher standard deviation indicates that the data points are more spread out from the mean, and conversely smaller and more concentrated.

min    2.000000    The smallest non-null value in the column.

25%    7.750000    The value below which 25% of the data falls. This is also known as the 1st quartile (Q1). It provides an idea of the lower end of the data distribution.

50%    13.500000   The middle value of the data, also called the median. It splits the data into two equal parts: half the data points are below the value and half the data points are above the value.

75%    18.250000   75% of the data are below this value. It provides insight into the upper bounds of the data distribution.

max    23.000000   The largest non-null value in the column.


If colum is categorical like id, the output might look like this:

count          20   The total number of non-null (non-missing) values in the column.

unique         20   The number of unique values ​​in the column, that is, how many different distinct values ​​there are.

top       admiral  The most frequent value in the column

freq            1  The frequency of the most frequent value

Name: id, dtype: object

df.describe() collect count, mean, and std, based only on non-missing (non-null) values in each column.

# 7.1
#Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']

df.dropna() deletes rows containing NULL values, del df['col'] deletes entire columns. For example, in a mall, some customers did not provide their email addresses, so the email column had missing values (NaN). But you still need basic information about other customers, such as names and phone numbers, that will be useful to you.
So you should use df.dropna(). del df[col] will delete all email addresses, but you still need the other customers' email addresses, so df.dropna[] will only delete those customers whose email addresses are missing.

# 7.2
#Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()

For example, when you ask customers to fill in information in a shopping mall, a few people fill in their gender. You realize that the information in the gender column is of no use to you because you're focused on their ID, name, and email. So you should use del df['col'] to delete the gender column. Instead of deleting rows for customers whose gender is missing

# 7.3
#Discuss why applying del df['col'] before df.dropna() when both are used together could be important

When you use df.dropna(), it drops any rows with missing values (NaN). If you use df.dropna() before del df['col'] you may lose rows containing valuable data just because that row has a NaN value.


# 7.4
#Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

Because there were multiple sets of data missing "gender", I thought it was no longer useful to me. I first used del df['gender'] to delete the gender column, and then I used df.dropna() to delete the rows that were missing other important information. Because these are useful to me.!


In [10]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

grouped_stats = df.groupby('pclass')['age'].describe()

print(grouped_stats)

        count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0


# 8.1
#Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you

df.groupby('pclass'): Group passengers by cabin class (pclass). In the Titanic dataset, passengers are divided into 3 categories: 1 = first class 2 = second class 3 = third class

['age']: After grouping by passenger class, select the age column,
.describe(): The age of each passenger is analyzed. 

count: The number of non-empty age values in each class.
mean: The average age of passengers in each class of travel.
std: standard deviation (a measure of age distribution).
min: The smallest passenger in each class.
25%, 50%, 75%: Percentiles, showing the age distribution within the class.
max: The oldest passenger in each class.



# 8.2 
#Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?

df.describe(): Collect the total number of non-null values in each column in the entire Data.
df.groupby("col1")["col2"].describe(): Collects the group-wise count of non-null values of col2 within each unique group of col1.

df.describe(): Simply ignore missing values ​​column-by-column. In df.groupby(): Ignore missing values for each specific group, which may produce different counts depending on how many missing values are in each group.


# 8.3 A

In [11]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

grouped_stats = df.groupby('pclass')['age'].describe()

print(grouped_stats)

        count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0


Need to import pandas as pd  # Make sure pandas is imported  

# 8.3B

In [12]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

grouped_stats = df.groupby('pclass')['age'].describe()

print(grouped_stats)

HTTPError: HTTP Error 404: Not Found

Here are some common reasons why an HTTPError might be raised:

Invalid or malformed request URL.

I accidentally added an extra "s" at the end of "titanic" in the URL. Here's the corrected code:

# 8.3C

In [13]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

grouped_stats = DF.groupby("col1")["col2"].describe()

print(grouped_stats)

NameError: name 'DF' is not defined

In [14]:
Variable names are case-sensitive, I used DF instead of df. The correct variable name is df

SyntaxError: invalid syntax (1038315438.py, line 1)

# 8.3D

In [15]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url

grouped_stats = dF.groupby("col1")["col2"].describe()

print(grouped_stats)

SyntaxError: '(' was never closed (3555961187.py, line 4)

In [16]:
I fogot the ) after pd.read_csv() and it was never close 

SyntaxError: unmatched ')' (1682438677.py, line 1)

# 8.3E

In [17]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

grouped_stats = df.group_by("col1")["col2"].describe()

print(grouped_stats)

AttributeError: 'DataFrame' object has no attribute 'group_by'

In [18]:
DataFrame do not have attribute group_by. The correct syntax for grouping data in pandas is groupby, not group_by.

SyntaxError: invalid syntax (1832045322.py, line 1)

# 8.3F

In [19]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

grouped_stats = df.groupby("Sex")["age"].describe()

print(grouped_stats)

KeyError: 'Sex'

In [ ]:
There is a problem with the name, sex should be lowercase instead of uppercase Sex

# 8.3G

In [20]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

grouped_stats = df.groupby(pclass)['age'].describe()

print(grouped_stats)

NameError: name 'pclass' is not defined

"sex" must be enclosed in quotes because it is a string that refers to a column in the DataFrame, not a variable.

# 9
#9. Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

Yes, when I face a problem I will check the piazza if anyone else has the same problem and try to find a solution.

https://chatgpt.com/share/66e3a074-78fc-8000-99c1-88b326bd5393

https://chatgpt.com/share/66e3a425-0aa0-8000-8a86-1d92a34da949

https://chatgpt.com/share/66e3a8a8-1ae0-8000-a926-1163901622f7

https://docs.google.com/document/d/11nh349AA6TKGNDueDdGXRNYB7D9YwNQga7RQPJ5zqyU/edit?addon_store#heading=h.lolf6fvu1i40 